### PR TITLE
dev: add ENVOY_DOCKER_SKIP_PULL to skip docker pull

### PR DESCRIPTION
**Pull request by spacewander**
Signed-off-by: spacewander <spacewanderlzx@gmail.com>

Commit Message: dev: add ENVOY_DOCKER_SKIP_PULL to skip docker pull
Additional Description: it  saves lots of time when developing with run_envoy_docker.sh wrapper
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: ci/run_envoy_docker.sh [hunk 1]
```diff
@@ -80,7 +80,9 @@ if ! is_windows; then
     VOLUMES+=(-v "${SHARED_TMP_DIR}":"${SHARED_TMP_DIR}")
 fi
 
-time docker pull "${ENVOY_BUILD_IMAGE}"
+if [[ -z "${ENVOY_DOCKER_SKIP_PULL}" ]]; then
+    time docker pull "${ENVOY_BUILD_IMAGE}"
```
Comment (phlax) [inline ci/run_envoy_docker.sh:84]: we could probably get rid of the `time` here also now i think

**Review by phlax — COMMENTED** [review]
happy to land this - separating the pull part was mostly done so we could see how much CI time was being used to pull the image

im wondering if we should make it the default to not pull separately

File: ci/run_envoy_docker.sh (hunk 1 above)
Comment (phlax) [reply ci/run_envoy_docker.sh:84]: altho - if we make it the non-default then perhaps it would be useful to keep

**Comment by spacewander** [comment]
> happy to land this - separating the pull part was mostly done so we could see how much CI time was being used to pull the image
> 
> im wondering if we should make it the default to not pull separately

If we make it the default, we need to update CI to always pull the latest one, otherwise a stale image may be used.

**Comment by phlax** [comment]
> ... we need to update CI to always pull the latest one ...

i think its ok - we dont need to worry about that so much as the build image uses a "unique" hash for its tag (of course it could theoretically change but wont unless something has gone wrong)

**Comment by spacewander** [comment]
Should I update the PR to make the docker pull non-default now, or should we wait for more opinions from the other people?

File: ci/README.md [hunk 2]
```diff
@@ -70,6 +70,12 @@ Besides `http_proxy` and `https_proxy`, maybe you need to set `go_proxy` to repl
 IMAGE_NAME=envoyproxy/envoy-build-ubuntu go_proxy=https://goproxy.cn,direct http_proxy=http://proxy.foo.com:8080 https_proxy=http://proxy.bar.com:8080 ./ci/run_envoy_docker.sh <build_script_args>
 ```
 
+The script will always pull the docker image. If you want to skip this action, you can specify `ENVOY_DOCKER_SKIP_PULL=true`.
+
+```bash
+ENVOY_DOCKER_SKIP_PULL=true ./ci/run_envoy_docker.sh <build_script_args>
```
Comment (phlax) [inline ci/README.md:76]:
```suggestion
ENVOY_DOCKER_PULL=true ./ci/run_envoy_docker.sh <build_script_args>
```

File: ci/run_envoy_docker.sh [hunk 3]
```diff
@@ -80,7 +80,9 @@ if ! is_windows; then
     VOLUMES+=(-v "${SHARED_TMP_DIR}":"${SHARED_TMP_DIR}")
 fi
 
-time docker pull "${ENVOY_BUILD_IMAGE}"
+if [[ -z "${ENVOY_DOCKER_SKIP_PULL}" ]]; then
```
Comment (phlax) [inline ci/run_envoy_docker.sh:83]:
```suggestion
if [[ -n "${ENVOY_DOCKER_PULL}" ]]; then
```

**Review by phlax — COMMENTED** [review]
> should we wait for more opinions from the other people?

if you update, im happy to land

File: ci/README.md [hunk 4]
```diff
@@ -70,6 +70,12 @@ Besides `http_proxy` and `https_proxy`, maybe you need to set `go_proxy` to repl
 IMAGE_NAME=envoyproxy/envoy-build-ubuntu go_proxy=https://goproxy.cn,direct http_proxy=http://proxy.foo.com:8080 https_proxy=http://proxy.bar.com:8080 ./ci/run_envoy_docker.sh <build_script_args>
 ```
 
+The script will always pull the docker image. If you want to skip this action, you can specify `ENVOY_DOCKER_SKIP_PULL=true`.
```
Comment (phlax) [inline ci/README.md:73]:
```suggestion
To force the Envoy build image to be refreshed by Docker you can set `ENVOY_DOCKER_PULL=true`.
```

**Review by phlax — APPROVED** [review]
great! thanks for working on this @spacewander , lgtm

im thinking about following up with something to quieten the docker pull that happens when it runs in CI

**Pull request merged**